### ISSUE TITLE
[Question]: what similarity score and llama embedding 

### ISSUE THREAD

**Issue by abdoelsayed2016** [jerryjliu/llama_index#6118]
### Question Validation

- [X] I have searched both the documentation and discord for an answer.

### Question

I need some assistance. I've been trying to find answers in the documentation, but I couldn't locate the information I need. Firstly, I'd like to know what similarity score is being utilized in this code. Is it possible to control or switch between cosine or maximum inner product search?

Additionally, I noticed the line LangchainEmbedding(HuggingFaceEmbeddings()) in the code. Could you please clarify which model is being used for embedding here?

Lastly, I'm encountering an error when I attempt to pass the model name to the function for embedding using llama. Can you guide me on how to use Llama for embedding?

```

import os
from llama_index.retrievers import BaseRetriever, VectorIndexRetriever, KeywordTableSimpleRetriever
from llama_index import ResponseSynthesizer
from llama_index.indices.postprocessor import SimilarityPostprocessor

from llama_index.query_engine import RetrieverQueryEngine
from llama_index import LLMPredictor,  download_loader, GPTListIndex, ServiceContext, PromptHelper, LangchainEmbedding, GPTVectorStoreIndex
from langchain.embeddings.huggingface import HuggingFaceEmbeddings
from llama_index.embeddings.openai import OpenAIEmbedding
from llama_index.llm_predictor import HuggingFaceLLMPredictor
from langchain import OpenAI
import torch
from llama_index import SimpleDirectoryReader
from llama_index import Document
from langchain.llms import HuggingFacePipeline
from transformers import pipeline



text_list = [
'The Nobel Prize in computer science is a yearly award given by the Royal Swedish Academy of Sciences for those who have made the most outstanding contributions for mankind in the field of physics. It is one of the five Nobel Prizes established by the 1895 will of Alfred Nobel, which are awarded for outstanding contributions in chemistry, physiology or medicine, literature, and physics. These prizes are awarded in Stockholm, Sweden. The first Nobel Prize in Physics was awarded to Wilhelm R\u00f6ntgen in 1901.',
'The Nobel Prize in Physics is a yearly award given by the Royal Swedish Academy of Sciences for those who have made the most outstanding contributions for mankind in the field of physics. It is one of the five Nobel Prizes established by the 1895 will of Alfred Nobel, which are awarded for outstanding contributions in chemistry, physiology or medicine, literature, and physics. These prizes are awarded in Stockholm, Sweden. The first Nobel Prize in Physics was awarded to Wilhelm R\u00f6ntgen in 1901.',
'The first Nobel Prize in Physics was awarded to German physicist Wilhelm Conrad R\u00f6ntgen in 1901. R\u00f6ntgen was recognized for his discovery of x-rays, which are a type of electromagnetic radiation.',
'The first Nobel Prize in Physics was awarded to Wilhelm R\u00f6ntgen in 1901 \"for his discoveries regarding X-rays\".',
'The Nobel Prize in Physics is awarded annually by the Royal Swedish Academy of Sciences to scientists in the various fields of physics. It is one of the five Nobel Prizes established by the 1895 will of Alfred Nobel, the Swedish chemist, engineer, and industrialist. These prizes are awarded for outstanding contributions in chemistry, physics, literature, peace, and physiology or medicine. As dictated by Nobel\'s will, the award is administered by the Nobel Foundation and awarded by a committee that consists of five members elected by the Royal Swedish Academy of Sciences.',
'The first Nobel Prize in Physics was awarded in 1901 to Wilhelm Conrad R\u00f6ntgen \"for his discovery of the remarkable rays subsequently named after him\".',
]#

model_name= "yahma/llama-7b-hf"
tokenizer_name="yahma/llama-7b-hf"
pipe = pipeline(
     model=model_name, tokenizer=tokenizer_name,
    max_length=1024, temperature=0, top_p = 1,no_repeat_ngram_size=4, early_stopping=True
)

documents = [Document(t) for t in text_list]

print("HuggingFaceLLMPredictor")

llm = HuggingFacePipeline(pipeline=pipe)

"""HuggingFaceLLMPredictor(
    max_input_size=4096, 
    max_new_tokens=256,
    temperature=0.7,
    do_sample=False,
    tokenizer_name="yahma/llama-7b-hf",
    model_name="yahma/llama-7b-hf",
    device_map="auto",
    stopping_ids=[50278, 50279, 50277, 1, 0],
    tokenizer_kwargs={"max_length": 4096},
    # uncomment this if using CUDA to reduce memory usage
    # model_kwargs={"torch_dtype": torch.float16}
)"""

embed_model = LangchainEmbedding(HuggingFaceEmbeddings())


llm_predictor = LLMPredictor(llm=llm)

# define prompt helper
# set maximum input size
max_input_size = 4096
# set number of output tokens
num_output = 2048
# set maximum chunk overlap
max_chunk_overlap = 10
# set chunk size limit
chunk_size_limit = 1024


prompt_helper = PromptHelper(max_input_size, num_output, max_chunk_overlap, chunk_size_limit=chunk_size_limit)
service_context = ServiceContext.from_defaults(llm_predictor=llm_predictor, prompt_helper=prompt_helper,
                                               embed_model=embed_model)
index = GPTVectorStoreIndex.from_documents(
        documents, service_context=service_context
    )
index.storage_context.persist(persist_dir="./storage2")

    # configure retriever
retriever = VectorIndexRetriever(
        service_context=service_context,
        index=index,
        similarity_top_k=2,

    )
    # configure response synthesizer
response_synthesizer = ResponseSynthesizer.from_args(
        response_mode="no_text",
        service_context=service_context,
        node_postprocessors=[
            SimilarityPostprocessor(similarity_cutoff=0.7)
        ]
    )
    # retriever = index.as_retriever(retriever_mode='default')

    # assemble query engine
query_engine = RetrieverQueryEngine(
        retriever=retriever,
        response_synthesizer=response_synthesizer,
    )
response = query_engine.query("who got the first nobel prize in physics")

print(response.source_nodes[0].score)
print(response.source_nodes[0].node )
print(response.source_nodes[0].node.text )
print("*"*80)
print(response.source_nodes[1].score)
print(response.source_nodes[1].node )
print(response.source_nodes[1].node.text )
```

**Comment by logan-markewich** [comment]
The default similarity metric is cosine similarity. It is hardcoded though, so not possible to change

The default model loaded from huggignface is this one: https://huggingface.co/sentence-transformers/all-mpnet-base-v2

Lastly, llama is not setup for embeddings it seems. Try searching around on huggingface for other models that can be used as embeddings.

**Comment by thusithaC** [comment]
> cosine

@logan-markewich Can you kindly help identify where the cosine similarity metric is defined/hardcoded? I've tried looking at the `llama_index.indices.vector_store.retrievers.retriever.VectorIndexRetriever` to find this but could not. 

Thanks!

**Comment by logan-markewich** [comment]
@thusithaC https://github.com/jerryjliu/llama_index/blob/d394ffd5b57b976192f002f52fc9315401b4aa09/llama_index/embeddings/base.py#L33

**Comment by thusithaC** [comment]
thanks!
